File: py/327.py
from typing import List

class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        n = len(nums)
        sums = [0] * (n + 1)
        for i in range(n):
            sums[i + 1] = sums[i] + nums[i]
        return self.countWhileMergeSort(sums, 0, n + 1, lower, upper)

    def countWhileMergeSort(self, sums, start, end, lower, upper):
        if end - start <= 1:
            return 0
        mid = (start + end) // 2
        count = self.countWhileMergeSort(sums, start, mid, lower, upper) + \
                self.countWhileMergeSort(sums, mid, end, lower, upper)
        j, k, t = mid, mid, mid
        cache = [0] * (end - start)
        r = 0
        for i in range(start, mid):
            while k < end and sums[k] - sums[i] < lower:
                k += 1
            while j < end and sums[j] - sums[i] <= upper:
                j += 1
            while t < end and sums[t] < sums[i]:
                cache[r] = sums[t]
                r += 1
                t += 1
            cache[r] = sums[i]
            r += 1
            count += j - k
        for i in range(t - start):
            sums[start + i] = cache[i]
        return count

        

if __name__ == '__main__':
    sol = Solution()
    print(sol.countRangeSum([-2, 5, -1], -2, 2))

File: cpp/327.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        int n = nums.size();
        vector<long> prefix_sum(n + 1, 0);
        for (int i = 0; i < n; i++) {
            prefix_sum[i + 1] = prefix_sum[i] + nums[i];
        }
        return countWhileMergeSort(prefix_sum, 0, n + 1, lower, upper);
    }

    int countWhileMergeSort(vector<long>& prefix_sum, int start, int end, int lower, int upper) {
        if (end - start <= 1) {return 0;}
        int mid = start + (end - start) / 2;
        int count = countWhileMergeSort(prefix_sum, start, mid, lower, upper) + countWhileMergeSort(prefix_sum, mid, end, lower, upper);
        int j = mid, k = mid, t = mid;
        long cache[end - start];
        for (int i = start, r = 0; i < mid; ++i, ++r) {
            while(k < end && prefix_sum[k] - prefix_sum[i] < lower) ++k;
            while(j < end && prefix_sum[j] - prefix_sum[i] <= upper) ++j;
            count += j - k;
            while(t < end && prefix_sum[t] < prefix_sum[i]) cache[r++] = prefix_sum[t++];
            cache[r] = prefix_sum[i];
        }
        copy(cache, cache + (t - start), prefix_sum.begin() + start);
        return count;
    }
};

int main() {
    Solution sol;
    vector<int> nums{-2, 5, -1};
    cout << sol.countRangeSum(nums, -2, 2) << endl;
}